backend/src/main/java/fk/prof/backend/model/association/impl/ZookeeperBasedBackendAssociationStore.java

```diff
@@ -0,0 +1,358 @@
+                    }
+                  }
+                }
+              } else {
+                BackendDetail existingBackend = backendDetailLookup.get(existingBackendAssociation);
+                if (!existingBackend.isDefunct()) {
+                  /**
+                   * Backend is not defunct, proceed with returning this assignment
+                   * This is a double check, this condition is checked before acquiring the lock as well to avoid locking cost in majority of the cases
+                   */
+                  if (logger.isDebugEnabled()) {
+                    logger.debug(String.format("process_group=%s, existing backend=%s, available",
+                        ProtoUtil.processGroupCompactRepr(processGroup), existingBackend.getBackendIPAddress()));
+                  }
+                  future.complete(existingBackendAssociation);
+                } else {
+                  /**
+                   * Presently assigned backend is defunct, find an available backend and if found, de-associate current backend and assign the new one to process group
+                   * Defunct backend is not de-associated eagerly because if no available backend is found, its better to wait for current backend to come back alive
+                   */
+                  availableBackendsByPriority.remove(existingBackend);
+                  BackendDetail newBackend = getAvailableBackendFromPriorityQueue();
+                  try {
+                    if (newBackend == null) {
+                      logger.warn(String.format("Presently assigned backend=%s for process_group=%s is defunct but cannot find any available backend so keeping assignment unchanged",
+                          existingBackendAssociation, ProtoUtil.processGroupCompactRepr(processGroup)));
+                      future.complete(existingBackendAssociation);
+                    } else {
+                      try {
+                        deAssociateBackendWithProcessGroup(existingBackend, processGroup);
+                        associateBackendWithProcessGroup(newBackend, processGroup);
+                        if (logger.isDebugEnabled()) {
+                          logger.debug(String.format("process_group=%s, de-associating existing backend=%s, associating new backend=%s",
+                              ProtoUtil.processGroupCompactRepr(processGroup), existingBackend.getBackendIPAddress(), newBackend.getBackendIPAddress()));
+                        }
+                        future.complete(newBackend.getBackendIPAddress());
+                      } catch (Exception ex) {
+                        future.fail(String.format("Cannot persist association of backend=%s with process_group=%s in zookeeper",
+                            newBackend.getBackendIPAddress(), processGroup));
+                      }
+                    }
+                  } finally {
+                    if (newBackend != null) {
```

Comment:
This too can go under else, because there is nothing to offer in if case anyway.